core-bundle/contao/forms/FormAltcha.php

```diff
@@ -0,0 +1,148 @@
+<?php
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao;
+
+use Contao\CoreBundle\Controller\AltchaController;
+use Contao\CoreBundle\String\HtmlAttributes;
+
+class FormAltcha extends Widget
+{
+	/**
+	 * @var boolean
+	 */
+	protected $blnForAttribute = true;
+
+	/**
+	 * @var string
+	 */
+	protected $strTemplate = 'form_altcha';
+
+	/**
+	 * @var string
+	 */
+	protected $prefix = 'widget widget-altcha';
+
+	protected HtmlAttributes $altchaAttributes;
+
+	protected string $altchaAuto = '';
+
+	protected bool $altchaHideLogo = false;
+
+	protected bool $altchaHideFooter = false;
+
+	protected int $altchaMaxNumber = 10000000;
+
+	/**
+	 * Use the raw request data.
+	 *
+	 * @param array $arrAttributes An optional attributes array
+	 */
+	public function __construct($arrAttributes=null)
+	{
+		parent::__construct($arrAttributes);
+
+		$this->useRawRequestData = true;
+	}
+
+	/**
+	 * Return a parameter.
+	 *
+	 * @param string $strKey The parameter name
+	 *
+	 * @return mixed The parameter value
+	 */
+	public function __get($strKey)
+	{
+		if ('altchaAttributes' === $strKey)
+		{
+			return $this->getAltchaAttributes();
+		}
+
+		return parent::__get($strKey);
+	}
+
+	/**
+	 * Generate the widget and return it as string.
+	 *
+	 * @return string The widget markup
+	 */
+	public function generate(): string
+	{
+		return sprintf('<altcha-widget %s></altcha-widget>', $this->getAltchaAttributes());
+	}
+
+	/**
+	 * Parse the template file and return it as string.
+	 *
+	 * @param array $arrAttributes An optional attributes array
+	 *
+	 * @return string The template markup
+	 */
+	public function parse($arrAttributes = null): string
+	{
+		$request = $this->getContainer()->get('request_stack')->getCurrentRequest();
+
+		if ($request && $this->getContainer()->get('contao.routing.scope_matcher')->isBackendRequest($request))
+		{
+			$objTemplate = new BackendTemplate('be_wildcard');
+			$objTemplate->title = $this->label;
+
+			return $objTemplate->parse();
+		}
+
+		$this->altchaAttributes = $this->getAltchaAttributes();
+
+		// TODO: Use ResponseContext, once it supports appending to <head>
+		$GLOBALS['TL_HEAD'][] = '<script async defer src="' . $this->asset('js/altcha.min.js', 'contao-components/altcha') . '" type="module"></script>';
```

Comment:
We should add this via template so it can be adjusted easily. Just like we do in the accordion etc.

Reply:
Moreover, we need to add the CSP nonce here.

Reply:
A `nonce` is only required for inline scripts. This is an external script, so it should not be necessary.

Reply:
This feature is only available in the Twig templates. 🤷‍♂️

Reply:
> This feature is only available in the Twig templates. 🤷‍♂️

You can also write `$GLOBALS['TL_HEAD'][]` in PHP templates. I agree that this should be transferred to the template.

Reply:
Changed in f391bf99ff7c3239ac42d4c91d4a946461a5cfcb.